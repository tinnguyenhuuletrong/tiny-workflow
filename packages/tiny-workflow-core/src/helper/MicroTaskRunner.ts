export class MicroTaskRunner {
  private pendingTasks: Promise<any>[] = [];
  private resolved: PromiseSettledResult<any>[] = [];

  public addTask(
    task: Promise<any>,
    opts?: {
      runId: string;
    }
  ) {
    this.pendingTasks.push(task);
    return task;
  }

  reset() {
    this.pendingTasks.length = 0;
    this.resolved.length = 0;
  }

  stats() {
    return {
      total: this.pendingTasks.length,
      done: this.resolved.length,
    };
  }

  /** waiting for all pending tasks settled */
  async idle() {
    while (this.resolved.length != this.pendingTasks.length) {
      this.resolved = await Promise.allSettled(this.pendingTasks);
    }
  }
}
