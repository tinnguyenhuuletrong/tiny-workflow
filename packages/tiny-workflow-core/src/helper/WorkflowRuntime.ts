import { DurableState } from "../DurableState";
import type {
  Constructor,
  DurableStateSystemEntry,
  SnapshotType,
} from "../type";

export type WorkflowRuntimeOpt<T extends DurableState> = {
  ctx: IRuntimeContext<T>;
  genRunId: () => Promise<string>;
  InstanceClass: Constructor<T>;
};

export interface IRuntimeContext<T extends DurableState> {
  attachRuntime(runtime: WorkflowRuntime<T>): void;

  save(runId: string, snapshotData: SnapshotType<T>): Promise<void>;
  load(runId: string): Promise<SnapshotType<T> | null>;

  lock(runId: string): Promise<void>;
  unlock(runId: string): Promise<void>;
}

export type WorkflowRunResult =
  | {
      runId: string;
      status: "need_resume";
      resumeEntry: DurableStateSystemEntry;
    }
  | {
      runId: string;
      status: "finished";
    }
  | {
      runId: string;
      status: "error";
      error?: any;
    };

export class WorkflowRuntime<T extends DurableState> {
  constructor(private opts: WorkflowRuntimeOpt<T>) {
    this.opts.ctx.attachRuntime(this);
  }

  get ctx() {
    return this.opts.ctx;
  }

  async run(ins: T, runId?: string): Promise<WorkflowRunResult> {
    if (!runId) runId = await this.opts.genRunId();

    try {
      await this.ctx.lock(runId);

      for await (const it of ins.exec(runId)) {
        if (it.resumeTrigger) {
          const resumeId = it.resumeTrigger.resumeId;
          const resumeEntry = ins.getResume(resumeId);
          if (!resumeEntry)
            throw new Error(`ResumeEntry not found resumeId=${resumeId}`);

          const data = ins.toJSON() as SnapshotType<T>;
          await this.opts.ctx.save(runId, data);

          return {
            runId,
            status: "need_resume",
            resumeEntry,
          };
        }
      }

      const data = ins.toJSON() as SnapshotType<T>;
      await this.opts.ctx.save(runId, data);
      return { runId, status: "finished" };
    } catch (error) {
      return { runId, status: "error", error };
    } finally {
      await this.ctx.unlock(runId);
    }
  }

  async resume(
    runId: string,
    {
      resumeId,
      resumePayload,
    }: {
      resumeId: string;
      resumePayload?: any;
    }
  ) {
    const ins = await this.createInstanceByRunId(
      runId,
      this.opts.InstanceClass
    );
    ins.resolveResume(resumeId, resumePayload);
    return this.run(ins, runId);
  }

  async createInstanceByRunId(runId: string, Cons: Constructor<T>) {
    const data = await this.opts.ctx.load(runId);
    if (!data) throw new Error(`snapshotData not found for runId=${runId}`);
    const ins = DurableState.fromJSON(Cons, data) as T;
    return ins;
  }
}
