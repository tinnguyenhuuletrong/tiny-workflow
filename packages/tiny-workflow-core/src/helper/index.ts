export * from "./MicroTaskRunner";
export * from "./SimpleContext";
export * from "./WorkflowRuntime";
