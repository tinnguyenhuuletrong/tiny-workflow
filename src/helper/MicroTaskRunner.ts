export class MicroTaskRunner {
  private pendingTasks: Promise<any>[] = [];
  private resolved: PromiseSettledResult<any>[] = [];

  public addTask(task: Promise<any>) {
    this.pendingTasks.push(task);
  }

  reset() {
    this.pendingTasks.length = 0;
    this.resolved.length = 0;
  }

  stats() {
    return {
      total: this.pendingTasks.length,
      done: this.resolved.length,
    };
  }

  async idle() {
    while (this.resolved.length != this.pendingTasks.length) {
      this.resolved = await Promise.allSettled(this.pendingTasks);
    }
  }
}
