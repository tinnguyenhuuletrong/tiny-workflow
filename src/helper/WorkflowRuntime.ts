import { DurableState } from "../DurableState";
import type { Constructor, DurableStateSystemEntry } from "../type";

export type SnapshotType<T extends DurableState> = ReturnType<T["toJSON"]>;
export type WorkflowRuntimeOpt<T extends DurableState> = {
  storage: IStorage<T>;
  genRunId: () => Promise<string>;
};

export interface IStorage<T extends DurableState> {
  save(runId: string, snapshotData: SnapshotType<T>): Promise<void>;
  load(runId: string): Promise<SnapshotType<T> | null>;
}

type WorkflowRunResult =
  | {
      runId: string;
      status: "need_resume";
      resumeEntry: DurableStateSystemEntry;
    }
  | {
      runId: string;
      status: "finished";
    };

export class WorkflowRuntime<T extends DurableState> {
  constructor(private opts: WorkflowRuntimeOpt<T>) {}

  async run(ins: T, runId?: string): Promise<WorkflowRunResult> {
    if (!runId) runId = await this.opts.genRunId();

    for await (const it of ins.exec(runId)) {
      if (it.resumeTrigger) {
        const resumeId = it.resumeTrigger.resumeId;
        const resumeEntry = ins.getResume(resumeId);
        if (!resumeEntry)
          throw new Error(`ResumeEntry not found resumeId=${resumeId}`);

        const data = ins.toJSON() as SnapshotType<T>;
        await this.opts.storage.save(runId, data);

        return {
          runId,
          status: "need_resume",
          resumeEntry,
        };
      }
    }

    const data = ins.toJSON() as SnapshotType<T>;
    await this.opts.storage.save(runId, data);
    return { runId, status: "finished" };
  }

  async resume(
    runId: string,
    Cons: Constructor<T>,
    {
      resumeId,
      resumePayload,
    }: {
      resumeId: string;
      resumePayload?: any;
    }
  ) {
    const ins = await this.getInsByRunId(runId, Cons);
    ins.resolveResume(resumeId, resumePayload);
    return this.run(ins, runId);
  }

  async getInsByRunId(runId: string, Cons: Constructor<T>) {
    const data = await this.opts.storage.load(runId);
    if (!data) throw new Error(`snapshotData not found for runId=${runId}`);
    const ins = DurableState.fromJSON(Cons, data) as T;
    return ins;
  }
}
