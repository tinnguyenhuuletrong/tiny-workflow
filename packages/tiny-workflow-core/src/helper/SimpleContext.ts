import { setTimeout as waitMs } from "node:timers/promises";
import type { DurableState } from "../DurableState";
import type { SnapshotType } from "../type";
import { MicroTaskRunner } from "./MicroTaskRunner";
import type {
  IRuntimeContext,
  WorkflowRuntime,
  WorkflowRunResult,
} from "./WorkflowRuntime";

export class MemoryLock {
  readonly promise: Promise<void>;
  private resolve!: Function;

  constructor() {
    this.promise = new Promise<void>((resolve) => {
      this.resolve = resolve;
    });
  }

  unlock() {
    this.resolve();
  }

  async wait() {
    return this.promise;
  }
}

export class SimpleContext<T extends DurableState>
  implements IRuntimeContext<T>
{
  readonly runner = new MicroTaskRunner();
  runtime!: WorkflowRuntime<T>;

  db = new Map<string, SnapshotType<T>>();
  dbLock = new Map<string, MemoryLock>();

  attachRuntime(runtime: WorkflowRuntime<T>) {
    if (this.runtime)
      throw new Error("can not re-attach. Something wrong here");
    this.runtime = runtime;
  }

  async lock(runId: string) {
    const tmp = this.dbLock.get(runId);
    if (tmp) await tmp.wait();

    const ins = new MemoryLock();
    this.dbLock.set(runId, ins);
  }

  async unlock(runId: string) {
    const tmp = this.dbLock.get(runId);
    if (tmp) {
      tmp.unlock();
      this.dbLock.delete(runId);
    }
  }

  async save(runId: string, snapshotData: SnapshotType<T>) {
    this.db.set(runId, snapshotData);
  }

  async load(runId: string) {
    const tmp = this.db.get(runId);
    if (!tmp) return null;
    return tmp;
  }

  async start() {
    this.runner.reset();
  }

  async shutdown() {
    for (const it of this.dbLock.entries()) {
      it[1].unlock();
    }
  }

  async withTransaction(runId: string, action: () => Promise<void>) {
    try {
      await this.lock(runId);
      await action();
    } catch (error) {
      throw error;
    } finally {
      await this.unlock(runId);
    }
  }

  async scheduleNextRun(runId: string, result: WorkflowRunResult) {
    switch (result.status) {
      case "need_resume":
        {
          switch (result.resumeEntry.type) {
            case "timer":
              {
                await waitMs(result.resumeEntry.resumeAfter - Date.now());
                const res = await this.runtime.resume(runId, {
                  resumeId: result.resumeEntry.resumeId,
                  resumePayload: undefined,
                });

                this.runner.addTask(this.scheduleNextRun(runId, res), {
                  runId,
                });
              }
              break;

            default:
              throw new Error(
                `can not handle resumeType=${result.resumeEntry.type}`
              );
          }
        }
        break;

      case "error": {
        console.error(`error runId=${runId} - detail=`, result);
        return;
      }
      default:
        break;
    }
  }
}
