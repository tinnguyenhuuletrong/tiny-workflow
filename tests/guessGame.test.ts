import { test, expect, mock } from "bun:test";
import {
  DurableState,
  type DurableStateIterator,
  type DurableStateReturn,
  type StepHandler,
  type StepIt,
} from "../src";
import * as TestHelper from "./helper";

enum EStep {
  step_begin = "step_begin",
  step_play = "step_play",
  step_end = "step_end",
}
type TStateShape = {
  number: number;
  answer?: number[];
};
type EAuditLog = "log";

class GuessGameState extends DurableState<EStep, TStateShape, EAuditLog> {
  // Type Safeguard & Auto register
  private _static(key: EStep): StepHandler<EStep> {
    return this[key];
  }
  private _collectAndRegisterSteps() {
    Object.values(EStep).map((step) =>
      this.stepHandler.set(step, this._static(step).bind(this))
    );
  }
  constructor() {
    super(EStep.step_begin, {
      withAuditLog: true,
      debug: true,
    });
    this._collectAndRegisterSteps();
  }

  private async *step_begin(): StepIt<EStep> {
    return { nextStep: EStep.step_play };
  }

  private async *step_play(): StepIt<EStep> {
    let count = 0;
    let question = `number between 0-100`;
    let answer = [];
    const guessNum = this.state.number;
    while (true) {
      const tmp = this.waitForEvent(`event_ask_${count++}`, {
        question,
      });
      if (tmp.it) yield tmp.it;
      const last_answer = tmp.value();
      answer.push(last_answer);

      if (last_answer === guessNum) {
        break;
      } else if (last_answer < guessNum) {
        question = `${last_answer} lesser than the answer - number between 0-100`;
      } else {
        question = `${last_answer} gretter than the answer - number between 0-100`;
      }
    }
    this.state.answer = answer;

    return { nextStep: EStep.step_end };
  }

  private async *step_end(): StepIt<EStep> {
    return { nextStep: null };
  }
}

test("guess_game_one_run", async () => {
  const ins = DurableState.fromJSON<
    EStep,
    TStateShape,
    EAuditLog,
    GuessGameState
  >(GuessGameState, {
    step: EStep.step_begin,
    state: {
      number: 15,
    },
  });

  TestHelper.mockResumeIdSeq(ins);
  const it = ins.exec();
  let tmp;

  tmp = await testHelperRunUntilResume({ it });
  testHelperResolveEvent("round_1", ins, tmp, 5);
  tmp = await testHelperRunUntilResume({ it });
  testHelperResolveEvent("round_2", ins, tmp, 20);
  tmp = await testHelperRunUntilResume({ it });
  testHelperResolveEvent("round_3", ins, tmp, 10);
  tmp = await testHelperRunUntilResume({ it });
  testHelperResolveEvent("round_4", ins, tmp, 15);
  tmp = await testHelperRunUntilResume({ it });
  expect(tmp.done).toBe(true);

  // snapshot check
  let snapshot = TestHelper.removeProps(ins.toJSON(), [
    "_at",
    "resumeAt",
    "resumeAfter",
  ]);
  expect(snapshot).toMatchSnapshot("guess_game_one_run");
});

test("guess_game_save_load_run", async () => {
  let ins = DurableState.fromJSON<
    EStep,
    TStateShape,
    EAuditLog,
    GuessGameState
  >(GuessGameState, {
    step: EStep.step_begin,
    state: {
      number: 95,
    },
  });

  let mockedResumeId = 0;
  TestHelper.mockResumeIdSeq(
    ins,
    mockedResumeId,
    (val: number) => (mockedResumeId = val)
  );
  let it = ins.exec();
  let tmp;

  async function simulateSaveLoad() {
    if (it) {
      await it.return(null);
      console.log("destroy");
    }

    console.log("re-create");
    ins = TestHelper.simulateSaveAndLoad(GuessGameState, ins);
    TestHelper.mockResumeIdSeq(
      ins,
      mockedResumeId,
      (val: number) => (mockedResumeId = val)
    );
    it = ins.exec();
  }

  let ctx = { it, maxIter: Number.MAX_VALUE, actualRunIter: 0 };
  tmp = await testHelperRunUntilResume(ctx);
  testHelperResolveEvent("round_1", ins, tmp, 5);

  await simulateSaveLoad();

  ctx = { it, maxIter: Number.MAX_VALUE, actualRunIter: 0 };
  tmp = await testHelperRunUntilResume(ctx);
  testHelperResolveEvent("round_2", ins, tmp, 45);

  await simulateSaveLoad();

  ctx = { it, maxIter: Number.MAX_VALUE, actualRunIter: 0 };
  tmp = await testHelperRunUntilResume(ctx);
  testHelperResolveEvent("round_3", ins, tmp, 75);
  expect(ctx.actualRunIter).toMatchSnapshot("round_3_runinfo");

  await simulateSaveLoad();

  ctx = { it, maxIter: Number.MAX_VALUE, actualRunIter: 0 };
  tmp = await testHelperRunUntilResume(ctx);
  testHelperResolveEvent("round_4", ins, tmp, 99);
  expect(ctx.actualRunIter).toMatchSnapshot("round_4_runinfo");

  await simulateSaveLoad();

  ctx = { it, maxIter: Number.MAX_VALUE, actualRunIter: 0 };
  tmp = await testHelperRunUntilResume(ctx);
  testHelperResolveEvent("round_5", ins, tmp, 95);
  expect(ctx.actualRunIter).toMatchSnapshot("round_5_runinfo");

  ctx = { it, maxIter: Number.MAX_VALUE, actualRunIter: 0 };
  tmp = await testHelperRunUntilResume(ctx);
  expect(tmp.done).toBe(true);

  // // snapshot check
  let snapshot = TestHelper.removeProps(ins.toJSON(), [
    "_at",
    "resumeAt",
    "resumeAfter",
  ]);
  expect(snapshot).toMatchSnapshot("guess_game_save_load_run");
});

// --------------------------------------------------------------------
//  Helper
// --------------------------------------------------------------------

async function testHelperRunUntilResume(ctx: {
  it: AsyncGenerator<
    DurableStateIterator<EStep>,
    DurableStateReturn<TStateShape> | null,
    unknown
  >;
  maxIter?: number;
  actualRunIter?: number;
}) {
  if (!ctx.maxIter) ctx.maxIter = Number.MAX_VALUE;
  ctx.actualRunIter = 0;
  while (++ctx.actualRunIter < ctx.maxIter) {
    let tmp = await ctx.it.next();
    if (tmp.done) return tmp;
    if (!tmp.done && tmp.value.resumeTrigger) return tmp;
  }

  throw new Error("OoO. Can not finished within maxRounds");
}

function testHelperResolveEvent(
  hint: string,
  ins: GuessGameState,
  tmp: IteratorResult<
    DurableStateIterator<EStep>,
    DurableStateReturn<TStateShape> | null
  >,
  answer: number
) {
  if (tmp.done) {
    throw new Error("OoO");
  }
  if (!tmp.value.resumeTrigger) throw new Error("Missing resume trigger");
  const resumeId = tmp.value.resumeTrigger.resumeId;
  ins.resolveResume(resumeId, answer);
  expect(ins.getResume(resumeId)).toMatchSnapshot(hint);
}
