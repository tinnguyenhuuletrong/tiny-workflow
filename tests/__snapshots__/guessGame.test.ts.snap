// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`guess_game_one_run: round_1 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "number between 0-100",
  },
  "responsePayload": 5,
  "resumeId": "event_ask_0-0",
  "type": "event",
}
`;

exports[`guess_game_one_run: round_2 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "5 lesser than the answer - number between 0-100",
  },
  "responsePayload": 20,
  "resumeId": "event_ask_1-1",
  "type": "event",
}
`;

exports[`guess_game_one_run: round_3 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "20 gretter than the answer - number between 0-100",
  },
  "responsePayload": 10,
  "resumeId": "event_ask_2-2",
  "type": "event",
}
`;

exports[`guess_game_one_run: round_4 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "10 lesser than the answer - number between 0-100",
  },
  "responsePayload": 15,
  "resumeId": "event_ask_3-3",
  "type": "event",
}
`;

exports[`guess_game_one_run: guess_game_one_run 1`] = `
{
  "cache": {
    "_seq": 2,
  },
  "logs": [
    {
      "_step": "step_play",
      "type": "transition",
      "values": {
        "from": "step_begin",
        "stepSeq": 1,
        "to": "step_play",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_0-0",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_0-0",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_1-1",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_1-1",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_2-2",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_2-2",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_3-3",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_3-3",
      },
    },
    {
      "_step": "step_end",
      "type": "transition",
      "values": {
        "from": "step_play",
        "stepSeq": 2,
        "to": "step_end",
      },
    },
  ],
  "state": {
    "answer": [
      5,
      20,
      10,
      15,
    ],
    "number": 15,
  },
  "step": "step_end",
  "system": {
    "1:step_play:event:event_ask_0": {
      "isDone": true,
      "requestPayload": {
        "question": "number between 0-100",
      },
      "responsePayload": 5,
      "resumeId": "event_ask_0-0",
      "type": "event",
    },
    "1:step_play:event:event_ask_1": {
      "isDone": true,
      "requestPayload": {
        "question": "5 lesser than the answer - number between 0-100",
      },
      "responsePayload": 20,
      "resumeId": "event_ask_1-1",
      "type": "event",
    },
    "1:step_play:event:event_ask_2": {
      "isDone": true,
      "requestPayload": {
        "question": "20 gretter than the answer - number between 0-100",
      },
      "responsePayload": 10,
      "resumeId": "event_ask_2-2",
      "type": "event",
    },
    "1:step_play:event:event_ask_3": {
      "isDone": true,
      "requestPayload": {
        "question": "10 lesser than the answer - number between 0-100",
      },
      "responsePayload": 15,
      "resumeId": "event_ask_3-3",
      "type": "event",
    },
  },
}
`;

exports[`guess_game_save_load_run: round_1 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "number between 0-100",
  },
  "responsePayload": 5,
  "resumeId": "event_ask_0-0",
  "type": "event",
}
`;

exports[`guess_game_save_load_run: round_2 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "5 lesser than the answer - number between 0-100",
  },
  "responsePayload": 45,
  "resumeId": "event_ask_1-1",
  "type": "event",
}
`;

exports[`guess_game_save_load_run: round_3 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "45 lesser than the answer - number between 0-100",
  },
  "responsePayload": 75,
  "resumeId": "event_ask_2-2",
  "type": "event",
}
`;

exports[`guess_game_save_load_run: round_3_runinfo 1`] = `1`;

exports[`guess_game_save_load_run: round_4 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "75 lesser than the answer - number between 0-100",
  },
  "responsePayload": 99,
  "resumeId": "event_ask_3-3",
  "type": "event",
}
`;

exports[`guess_game_save_load_run: round_4_runinfo 1`] = `1`;

exports[`guess_game_save_load_run: round_5 1`] = `
{
  "isDone": true,
  "requestPayload": {
    "question": "99 gretter than the answer - number between 0-100",
  },
  "responsePayload": 95,
  "resumeId": "event_ask_4-4",
  "type": "event",
}
`;

exports[`guess_game_save_load_run: round_5_runinfo 1`] = `1`;

exports[`guess_game_save_load_run: guess_game_save_load_run 1`] = `
{
  "cache": {
    "_seq": 2,
  },
  "logs": [
    {
      "_step": "step_play",
      "type": "transition",
      "values": {
        "from": "step_begin",
        "stepSeq": 1,
        "to": "step_play",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_0-0",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_0-0",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_1-1",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_1-1",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_2-2",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_2-2",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_3-3",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_3-3",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_begin",
      "values": {
        "resumeId": "event_ask_4-4",
        "type": "event",
      },
    },
    {
      "_step": "step_play",
      "type": "interrupt_end",
      "values": {
        "resumeId": "event_ask_4-4",
      },
    },
    {
      "_step": "step_end",
      "type": "transition",
      "values": {
        "from": "step_play",
        "stepSeq": 2,
        "to": "step_end",
      },
    },
  ],
  "state": {
    "answer": [
      5,
      45,
      75,
      99,
      95,
    ],
    "number": 95,
  },
  "step": "step_end",
  "system": {
    "1:step_play:event:event_ask_0": {
      "isDone": true,
      "requestPayload": {
        "question": "number between 0-100",
      },
      "responsePayload": 5,
      "resumeId": "event_ask_0-0",
      "type": "event",
    },
    "1:step_play:event:event_ask_1": {
      "isDone": true,
      "requestPayload": {
        "question": "5 lesser than the answer - number between 0-100",
      },
      "responsePayload": 45,
      "resumeId": "event_ask_1-1",
      "type": "event",
    },
    "1:step_play:event:event_ask_2": {
      "isDone": true,
      "requestPayload": {
        "question": "45 lesser than the answer - number between 0-100",
      },
      "responsePayload": 75,
      "resumeId": "event_ask_2-2",
      "type": "event",
    },
    "1:step_play:event:event_ask_3": {
      "isDone": true,
      "requestPayload": {
        "question": "75 lesser than the answer - number between 0-100",
      },
      "responsePayload": 99,
      "resumeId": "event_ask_3-3",
      "type": "event",
    },
    "1:step_play:event:event_ask_4": {
      "isDone": true,
      "requestPayload": {
        "question": "99 gretter than the answer - number between 0-100",
      },
      "responsePayload": 95,
      "resumeId": "event_ask_4-4",
      "type": "event",
    },
  },
}
`;
